File: src/server/aggregators/reddit/urls.ts
/**
 * Reddit URL utilities.
 */

import axios from "axios";

import { logger } from "@server/utils/logger";

import { getRedditAccessToken } from "./auth";

/**
 * Reddit subreddit info interface.
 */
interface RedditSubredditInfo {
  data: {
    display_name: string;
    icon_img: string;
    community_icon: string;
    header_img: string | null;
  };
}

/**
 * Decode HTML entities in URLs.
 * Converts &amp; to &, &lt; to <, &gt; to >, &quot; to ", &#39; to '.
 */
export function decodeHtmlEntitiesInUrl(url: string): string {
  return url
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

/**
 * Fix redditmedia.com and external-preview.redd.it URLs by replacing &amp; with &.
 * Also decodes HTML entities in the URL first.
 */
export function fixRedditMediaUrl(url: string | null): string | null {
  if (!url) return null;
  // Decode HTML entities first
  const decoded = decodeHtmlEntitiesInUrl(url);
  if (
    decoded.includes("styles.redditmedia.com") ||
    decoded.includes("external-preview.redd.it")
  ) {
    return decoded.replace(/&amp;/g, "&");
  }
  return decoded;
}

/**
 * Convert Reddit preview.redd.it URLs to i.redd.it URLs when possible.
 * Reddit's i.redd.it CDN is more accessible than preview.redd.it.
 */
export function convertRedditPreviewUrl(url: string): string {
  try {
    // Convert preview.redd.it to i.redd.it
    if (url.includes("preview.redd.it")) {
      const urlObj = new URL(url);
      // Extract the filename from the path
      const pathParts = urlObj.pathname.split("/");
      const filename = pathParts[pathParts.length - 1];

      // Build i.redd.it URL (remove query params as they're often signatures)
      const newUrl = `https://i.redd.it/${filename}`;
      logger.debug(
        { original: url, converted: newUrl },
        "Converting Reddit preview URL",
      );
      return newUrl;
    }
    return url;
  } catch (error) {
    logger.debug({ error, url }, "Failed to convert Reddit preview URL");
    return url;
  }
}

/**
 * Get appropriate referer header for Reddit URLs.
 * For Reddit URLs, use reddit.com. For others, use the domain of the URL.
 */
export function getRedditRefererHeader(url: string): string {
  try {
    const urlObj = new URL(url);

    // Special handling for Reddit domains
    if (
      urlObj.hostname.includes("redd.it") ||
      urlObj.hostname.includes("reddit.com")
    ) {
      return "https://www.reddit.com";
    }

    // For other domains, use the origin
    return `${urlObj.protocol}//${urlObj.hostname}`;
  } catch (error) {
    logger.debug({ error, url }, "Failed to determine referer");
    return "https://www.reddit.com"; // Safe fallback for Reddit
  }
}

/**
 * Extract subreddit name from URL or identifier.
 */
export function normalizeSubreddit(identifier: string): string {
  identifier = identifier.trim();

  // Extract from URL
  const urlMatch = /(?:reddit\.com)?\/r\/(\w+)/.exec(identifier);
  if (urlMatch) {
    return urlMatch[1];
  }

  // Remove r/ or /r/ prefix
  if (identifier.startsWith("/r/")) {
    return identifier.slice(3);
  }
  if (identifier.startsWith("r/")) {
    return identifier.slice(2);
  }

  return identifier;
}

/**
 * Extract post ID and subreddit from Reddit URL.
 * Format: https://reddit.com/r/{subreddit}/comments/{postId}/...
 */
export function extractPostInfoFromUrl(url: string): {
  subreddit: string | null;
  postId: string | null;
} {
  const match = /\/r\/(\w+)\/comments\/([a-zA-Z0-9]+)/.exec(url);
  return match
    ? { subreddit: match[1], postId: match[2] }
    : { subreddit: null, postId: null };
}

/**
 * Validate subreddit name.
 */
export function validateSubreddit(subreddit: string): {
  valid: boolean;
  error?: string;
} {
  if (!subreddit) {
    return { valid: false, error: "Subreddit is required" };
  }

  // Subreddit names: 2-21 characters, alphanumeric and underscores only
  if (!/^\w{2,21}$/.test(subreddit)) {
    return {
      valid: false,
      error:
        "Invalid subreddit name. Use 2-21 alphanumeric characters or underscores.",
    };
  }

  return { valid: true };
}

/**
 * Fetch subreddit information including icon.
 */
export async function fetchSubredditInfo(
  subreddit: string,
  userId: number,
): Promise<{ iconUrl: string | null }> {
  try {
    const accessToken = await getRedditAccessToken(userId);
    const url = `https://oauth.reddit.com/r/${subreddit}/about`;
    const response = await axios.get<RedditSubredditInfo>(url, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      timeout: 10000,
    });

    const subredditData = response.data.data;
    // Prefer icon_img, fall back to community_icon
    // Decode HTML entities before fixing the URL
    const rawIconUrl =
      subredditData.icon_img || subredditData.community_icon || null;
    const iconUrl = rawIconUrl
      ? fixRedditMediaUrl(decodeHtmlEntitiesInUrl(rawIconUrl))
      : null;

    if (iconUrl) {
      logger.debug({ subreddit, iconUrl }, "Fetched subreddit icon");
    }

    return { iconUrl };
  } catch (error) {
    logger.warn({ error, subreddit }, "Failed to fetch subreddit info");
    return { iconUrl: null };
  }
}

/**
 * Extract URLs from Reddit post text (selftext).
 * Handles both plain URLs and markdown links [text](url).
 * Decodes HTML entities in extracted URLs.
 */
export function extractUrlsFromText(text: string): string[] {
  if (!text) return [];

  const urls: string[] = [];

  // Pattern for markdown links: [text](url)
  const markdownLinkPattern = /\[([^\]]*)\]\((https?:\/\/[^)]+)\)/g;
  let match;
  while ((match = markdownLinkPattern.exec(text)) !== null) {
    urls.push(decodeHtmlEntitiesInUrl(match[2]));
  }

  // Pattern for plain URLs: http:// or https://
  // This regex matches URLs but avoids matching URLs already found in markdown links
  const plainUrlPattern = /(?<!\]\()(https?:\/\/[^\s)]+)/g;
  while ((match = plainUrlPattern.exec(text)) !== null) {
    // Remove trailing punctuation that might be part of the sentence
    const url = match[1].replace(/[.,;:!?)]+$/, "");
    const decodedUrl = decodeHtmlEntitiesInUrl(url);
    if (!urls.includes(decodedUrl)) {
      urls.push(decodedUrl);
    }
  }

  return urls;
}

File: core/aggregators/reddit/urls.py
"""Reddit URL utilities."""

import logging
import re
from typing import Any, Dict, Optional

import requests

from .auth import get_reddit_access_token

logger = logging.getLogger(__name__)

REDDIT_API_BASE = "https://www.reddit.com"


def decode_html_entities_in_url(url: str) -> str:
    """
    Decode HTML entities in URLs.

    Converts &amp; to &, &lt; to <, &gt; to >, &quot; to ", &#39; to '.

    Args:
        url: URL string

    Returns:
        Decoded URL
    """
    return (
        url.replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", '"')
        .replace("&#39;", "'")
    )


def fix_reddit_media_url(url: Optional[str]) -> Optional[str]:
    """
    Fix redditmedia.com and external-preview.redd.it URLs.

    Replaces &amp; with & and decodes HTML entities.

    Args:
        url: URL from Reddit API

    Returns:
        Fixed URL or None
    """
    if not url:
        return None

    decoded = decode_html_entities_in_url(url)
    if "styles.redditmedia.com" in decoded or "external-preview.redd.it" in decoded:
        return decoded.replace("&amp;", "&")
    return decoded


def normalize_subreddit(identifier: str) -> str:
    """
    Extract subreddit name from URL or identifier.

    Args:
        identifier: Subreddit identifier (can be URL, r/subreddit, or just subreddit)

    Returns:
        Normalized subreddit name
    """
    identifier = identifier.strip()

    # Extract from URL
    url_match = re.search(r"(?:reddit\.com)?/r/(\w+)", identifier)
    if url_match:
        return url_match.group(1)

    # Remove r/ or /r/ prefix
    if identifier.startswith("/r/"):
        return identifier[3:]
    if identifier.startswith("r/"):
        return identifier[2:]

    return identifier


def extract_post_info_from_url(url: str) -> Dict[str, Optional[str]]:
    """
    Extract post ID and subreddit from Reddit URL.

    Format: https://reddit.com/r/{subreddit}/comments/{postId}/...

    Args:
        url: Reddit post URL

    Returns:
        Dict with 'subreddit' and 'post_id' keys
    """
    match = re.search(r"/r/(\w+)/comments/([a-zA-Z0-9]+)", url)
    if match:
        return {"subreddit": match.group(1), "post_id": match.group(2)}
    return {"subreddit": None, "post_id": None}


def validate_subreddit(subreddit: str) -> Dict[str, Any]:
    """
    Validate subreddit name.

    Args:
        subreddit: Subreddit name

    Returns:
        Dict with 'valid' (bool) and optional 'error' (str)
    """
    if not subreddit:
        return {"valid": False, "error": "Subreddit is required"}

    # Subreddit names: 2-21 characters, alphanumeric and underscores only
    if not re.match(r"^\w{2,21}$", subreddit):
        return {
            "valid": False,
            "error": "Invalid subreddit name. Use 2-21 alphanumeric characters or underscores.",
        }

    return {"valid": True}


def fetch_subreddit_info(subreddit: str, user_id: int) -> Dict[str, Optional[str]]:
    """
    Fetch subreddit information including icon.

    Args:
        subreddit: Subreddit name (without /r/)
        user_id: User ID for authentication

    Returns:
        Dict with 'iconUrl' key
    """
    try:
        access_token = get_reddit_access_token(user_id)
        url = f"https://oauth.reddit.com/r/{subreddit}/about"
        response = requests.get(
            url,
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=10,
        )
        response.raise_for_status()

        data = response.json()
        subreddit_data = data.get("data", {})

        # Prefer icon_img, fall back to community_icon
        raw_icon_url = subreddit_data.get("icon_img") or subreddit_data.get("community_icon")
        icon_url = None
        if raw_icon_url:
            icon_url = fix_reddit_media_url(decode_html_entities_in_url(raw_icon_url))

        if icon_url:
            logger.debug(f"Fetched subreddit icon for r/{subreddit}: {icon_url}")

        return {"iconUrl": icon_url}

    except Exception as e:
        logger.warning(f"Failed to fetch subreddit info for r/{subreddit}: {e}")
        return {"iconUrl": None}


def extract_urls_from_text(text: str) -> list[str]:
    """
    Extract URLs from Reddit post text (selftext).

    Handles both plain URLs and markdown links [text](url).
    Decodes HTML entities in extracted URLs.

    Args:
        text: Text to extract URLs from

    Returns:
        List of URLs
    """
    if not text:
        return []

    urls = []

    # Pattern for markdown links: [text](url)
    markdown_link_pattern = re.compile(r"\[([^\]]*)\]\((https?://[^)]+)\)")
    for match in markdown_link_pattern.finditer(text):
        urls.append(decode_html_entities_in_url(match.group(2)))

    # Pattern for plain URLs: http:// or https://
    plain_url_pattern = re.compile(r"(?<!\]\()(https?://[^\s)]+)")
    for match in plain_url_pattern.finditer(text):
        # Remove trailing punctuation
        url = re.sub(r"[.,;:!?)]+$", "", match.group(1))
        decoded_url = decode_html_entities_in_url(url)
        if decoded_url not in urls:
            urls.append(decoded_url)

    return urls
